### DISCUSSION TITLE
Custom API Rules

### DISCUSSION THREAD

**Discussion by joaoseidel** [pocketbase/pocketbase#2607] [Q&A] [answered]
Let's say I have a `Books` collection with a field `status` that can have `ongoing`, `paused` or `archived` values. I have a specific rule for the `archived` status that only the owner of the book can have access to it, how can I achieve this behavior using PB API Rules? I couldn't do it normally, so I've tried using the SDK, but if I add a filter, let's say on a list endpoint, It messes with the page info response. Can anyone help me?

Thanks in advance

**Comment by ganigeorgiev** [comment]
> I have a specific rule for the archived status that only the owner of the book can have access to it, how can I achieve this behavior using PB API Rules?

I'm not sure that I understand what do you mean. Could elaborate and/or provide an example of the desired behavior?

In general, you can apply a conditional rule with `||` or `&&`, for example:

```js
(owner = @request.auth.id && status = "archived") || (...something else...)
```

**Reply by joaoseidel** [reply]
Yeah, of course I can provide you more details. Let's say I need to perform a GET on the List endpoint of Books collection, ok? So the reason why the way you did do not work, is that I need to filter every record and if one have the status archived, it can only be returned if the requester is the owner of the book, and if not, those records are not returned, this is mostly for security reasons when I don't want to expose private data for everyone else except the owner. On a search page for example, anyone can get the response of the fetch on network tab of the DevTools, so I need to treat this in the server side, not on the client side. I hope now it makes more clear

**Reply by ganigeorgiev** [reply]
@joaoseidel I'm sorry but I still don't understand why using the collection API rules wouldn't work in your case. You can think of the API rules as the `WHERE` condition in a SQL query. API rules are applied only server-side and they can't be modified client-side. Please provide a more concrete example.

**Reply by joaoseidel** [reply]
Let's try to exemplify what I need:

I have these users that want to GET /api/collections/books/records:
- Anonymous
- User A (id: izcg7xs9qqz0vjn)
- User B (id: 84ddu1lm0idhbws)

And I have these books records:
- Book 1 (id: ovrsz2wj9ki9a47, status: ONGOING, owner: izcg7xs9qqz0vjn)
- Book 2 (id: n43aepj7c45sqfc, status: ARCHIVED, owner: izcg7xs9qqz0vjn)
- Book 3 (id: 84ddu1lm0idhbws, status: ONGOING, owner: 349nhd7sp0dl50w)

Then, those users perform the GET without any API Rule, the response will be:
```json
{
  "page": 1,
  "perPage": 30,
  "totalItems": 3,
  "totalPages": 1,
  "items": [
    {
      "collectionId": "jcvhbopv1b55re5",
      "collectionName": "books",
      "created": "2023-05-31 20:43:52.762Z",
      "id": "ovrsz2wj9ki9a47",
      "owner": "izcg7xs9qqz0vjn",
      "status": "ONGOING",
      "updated": "2023-05-31 20:43:52.762Z"
    },
    {
      "collectionId": "jcvhbopv1b55re5",
      "collectionName": "books",
      "created": "2023-05-31 20:43:57.644Z",
      "id": "n43aepj7c45sqfc",
      "owner": "izcg7xs9qqz0vjn",
      "status": "ARCHIVED",
      "updated": "2023-05-31 20:43:57.644Z"
    },
    {
      "collectionId": "jcvhbopv1b55re5",
      "collectionName": "books",
      "created": "2023-05-31 20:44:06.679Z",
      "id": "84ddu1lm0idhbws",
      "owner": "349nhd7sp0dl50w",
      "status": "ONGOING",
      "updated": "2023-05-31 20:44:10.253Z"
    }
  ]
}
```

But what I want is that for **User B** and **Anonymous** the response should be:
```json
{
  "page": 1,
  "perPage": 30,
  "totalItems": 2,
  "totalPages": 1,
  "items": [
    {
      "collectionId": "jcvhbopv1b55re5",
      "collectionName": "books",
      "created": "2023-05-31 20:43:52.762Z",
      "id": "ovrsz2wj9ki9a47",
      "owner": "izcg7xs9qqz0vjn",
      "status": "ONGOING",
      "updated": "2023-05-31 20:43:52.762Z"
    },
    {
      "collectionId": "jcvhbopv1b55re5",
      "collectionName": "books",
      "created": "2023-05-31 20:44:06.679Z",
      "id": "84ddu1lm0idhbws",
      "owner": "349nhd7sp0dl50w",
      "status": "ONGOING",
      "updated": "2023-05-31 20:44:10.253Z"
    }
  ]
}
```

With a regular backend, I can simply achieve this behavior with a criteria appended to the query.

I managed to do it using the SDK, but it messes up the page info response, like totalItems instead of responding 2 items it would be responding 3 bc the filter was done after the query (using **OnRecordsListRequest** hook) 

There's a way of adding those filters using PB?

**Reply by ganigeorgiev** [reply, marked as answer]
I think you are confusing the client-side `filter` parameter with the [Collection API rules](https://pocketbase.io/docs/api-rules-and-filters/). They share the same syntax but the main difference is that API rules can be set only by the admins. The client-side `filter` defines extra condition to the API rules, it doesn't overwrite them.

Your example is still slightly confusing to me. I assume that the id of "User B" is a typo and should be `349nhd7sp0dl50w`, right?

I'll also assume that `owner` is marked "Nonempty".

If I'm understanding correctly, you want to:
- allow guests and non-owners to be able to list every book that is **not** `ARCHIVED`
- allow returning `ARCHIVED` books **only** if the current auth user is the owner of the archived book

If that's the case, then you can try the following Collection "List/Search API rule":

```js
status != "ARCHIVED" ||
(status = "ARCHIVED" && @request.auth.id = owner)

// or more simple

status != "ARCHIVED" || @request.auth.id = owner
```

**Reply by joaoseidel** [reply]
Omg how could I not do it myself, it's so simple! hahhaaha 

Thank you, this was exactly what I was looking for, and sorry for the misunderstanding (I'm non-native english speaker)

One last question: is it possible to create this exactly behavior, but per-field too?

**Reply by ganigeorgiev** [reply]
>  is it possible to create this exactly behavior, but per-field too?

No, the API rules apply on entire record(s).

Alternatively, you can create a helper [View collection](https://pocketbase.io/docs/collections/#view-collection) to assist you with hiding fields server-side and/or performing checks based on some aggregations (_to join and check another collection field you can use the `@collection.*` filter_).
